File: cross-account-publish/certificate-based-stack/python/publish.py
#!/usr/bin/env python3

import argparse

import requests
from OpenSSL import crypto

parser = argparse.ArgumentParser(
    description='Publish a message to IoT Core using the endpoint and topic information embedded in a certificate')
parser.add_argument('--cert', help='The file that contains the certificate', type=ascii)
parser.add_argument('--key',
                    help='The file that contains the private key. Optional if the certificate key\'s filename ends in \'.key\' and is otherwise the same as the certificate\'s filename (e.g. iot.pem and iot.key).',
                    type=ascii)
parser.add_argument('--message',
                    help='The message that should be published. This is a string and will sent exactly as specified. It is not automatically converted to JSON.',
                    type=ascii)

args = parser.parse_args()

if args.cert is None:
    raise Exception("The path to the certificate is a required argument")

if args.key is None:
    args.key = str(args.cert)
    extension_index = args.key.rindex(".")
    args.key = args.key[:extension_index] + ".key"

if args.message is None:
    raise Exception("The message is a required argument")

args.cert = args.cert.replace("'", "")
args.key = args.key.replace("'", "")

cert = crypto.load_certificate(crypto.FILETYPE_PEM, open(args.cert).read())

endpoint = cert.get_issuer().CN
topic = cert.get_subject().CN.replace("%2F", "/")

publish_url = 'https://' + endpoint + ':8443/topics/' + topic + '?qos=1'


def cross_account_publish(message):
    headers = {'x-amzn-platform': 'APN/1 partner,solution'}

    publish = requests.request('POST', publish_url, data=str.encode(message), cert=[args.cert, args.key],
                               headers=headers)


message = args.message

if len(message) > 2:
    if message[0] == '\'':
        message = message[1:]
    if message[len(message) - 1] == '\'':
        message = message[:-1]

cross_account_publish(message)

File: cross-account-publish/certificate-based-stack/nodejs/publish.js
#!/usr/bin/env node

const {X509Certificate} = require('crypto');
const {ArgumentParser} = require('argparse');
const FS = require('fs');
const https = require('https');

const parser = new ArgumentParser();

function getCommonName(input) {
    if (input.indexOf("CN=") === -1) {
        // No common name found
        return undefined;
    }

    return input.replace(/^.*CN=/, "").replace(/,.*$/, "")
}

parser.add_argument('-k', '--key', {help: 'The location of the private key file'});
parser.add_argument('-c', '--cert', {help: 'The location of the public certificate file'});
parser.add_argument('-m', '--message', {help: 'The message to publish'});

const arguments = parser.parse_args();

const keyFilename = arguments['key'];
const publicCertificateFilename = arguments['cert'];
const message = arguments['message'];

let argumentsSpecified = true;

if (typeof keyFilename === 'undefined') {
    argumentsSpecified = false;
    console.log('The private key filename was not specified')
}

if (typeof publicCertificateFilename === 'undefined') {
    argumentsSpecified = false;
    console.log('The public certificate filename was not specified')
}

if (typeof message === 'undefined') {
    argumentsSpecified = false;
    console.log('The message was not specified')
}

if (!argumentsSpecified) {
    process.exit(1)
}

let publicCertificateData = null;

try {
    publicCertificateData = FS.readFileSync(publicCertificateFilename, 'utf8');
} catch (e) {
    console.log('Failed to read public certificate file [', publicCertificateFilename, ']: ', e.stack);
    process.exit(1)
}

const publicCertificate = new X509Certificate(publicCertificateData);

const endpoint = getCommonName(publicCertificate.issuer)
const topic = getCommonName(publicCertificate.subject).replaceAll("%2F", "/")

if (typeof endpoint === 'undefined') {
    console.log('No endpoint information was found in the certificate\'s issuer common name field');
    process.exit(1)
}

if (typeof topic === 'undefined') {
    console.log('No topic information was found in the certificate\'s subject common name field');
    process.exit(1)
}

const options = {
    protocol: 'https:',
    hostname: endpoint,
    port: 8443,
    path: '/topics/' + topic + '?qos=1',
    method: 'POST',
    // headers: {
    //     'Content-Type': 'application/json',
    //     'Content-Length': data.length
    // },
    key: FS.readFileSync(keyFilename),
    cert: FS.readFileSync(publicCertificateFilename)//publicCertificateData
};

const req = https.request(options, (res) => {
    let data = '';

    res.on('data', (chunk) => {
        data += chunk;
    });

    res.on('end', () => {
        console.log(JSON.parse(data));
    });

}).on("error", (err) => {
    console.log("Error: ", err.message);
});

req.write(message);
req.end();
